sktime/utils/git_diff.py

```diff
@@ -78,3 +78,69 @@ class to get module string from, e.g., NaiveForecaster
     """
     module_str = get_module_from_class(cls)
     return is_module_changed(module_str)
+
+
+def get_changed_lines(file_path):
+    """Get changed or added lines from a file.
+
+    Parameters
+    ----------
+    file_path : str
+        path to file to get changed lines from
+
+    Returns
+    -------
+    list of str : changed lines
+    """
+    cmd = f"git diff remotes/origin/main -- {file_path}"
+
+    try:
+        # Run 'git diff' command to get the changes in the specified file
+        result = subprocess.check_output(cmd, shell=True, text=True)
+
+        # Extract the changed or new lines and return as a list of strings
+        changed_lines = [
+            line.strip() for line in result.split("\n") if line.startswith("+ ")
+        ]
+        # remove first character ('+') from each line
+        changed_lines = [line[1:] for line in changed_lines]
+
+        return changed_lines
+
+    except subprocess.CalledProcessError:
+        return []
+
+
+def get_packages_with_changed_specs():
+    """Get packages with changed or added specs.
+
+    Returns
+    -------
+    list of str : names of packages with changed or added specs
+    """
+    from packaging.requirements import Requirement
+
+    changed_lines = get_changed_lines("pyproject.toml")
```

Comment:
Comments:

1. is it relative with respect to repository root?
2. if two lines are switched in the file, for instance due to alphabetical sorting, will it be triggered?
3. What happens for formatting changes, e.g. space changes?
4. What happens for new releases when version line gets modified? Will it fail CI as that line changes can't be parsed by Requirement?
5. What happens when line is changed that does not specify version of a soft dependency?

1 is a clarification, 2-3 are minor doubts, 4-5 are main questions.

Reply:
> 1. is it relative with respect to repository root?

Yes, according to empirical testing that seems to be the case. Not sure how to ensure this - but it is similar to the existing `git_diff` utility, with `git` executed in the root.

> 2. if two lines are switched in the file, for instance due to alphabetical sorting, will it be triggered?

I thnk yes, and it should be, right?

> 3. What happens for formatting changes, e.g. space changes?

Also 

> 4. What happens for new releases when version line gets modified? Will it fail CI as that line changes can't be parsed by Requirement?

I think that's still going to be fine, since we are not using `Requirement` to find the impacted packages.

> 5. What happens when line is changed that does not specify version of a soft dependency?

Hmmmm - possibly an exception with current code? I hope I coded it robustly enough that nothing happens, but I will check.

Reply:
> Yes, according to empirical testing that seems to be the case. Not sure how to ensure this - but it is similar to the existing git_diff utility, with git executed in the root.

Okay, was just confirming. I think `git diff` then needs to be run from repository root, which is already happening as there is no `cd` step in CI, so it's ensured I think.

> I thnk yes, and it should be, right?

To save unnecessary runs, e.g. formatting/linting changes, it'd be better not to. But too tricky to detect, so I agree it's safer to run always.

---

I am still concerned with 4/5. Now a bit more, because it does not run correctly in my laptop.

This is what I did:

1. I changed line 4 of pyproject.toml by adding " data" at end.
2. I imported `get_changed_lines` and ran.
3. I received `[]`.

I debugged a bit by adding `print` statements, and it seems the reason is output of `git diff` seemed to start with just `+` and not `+ `. I don't know if it varies per operating system or git version or etc.

So, my new question is it always adding a space for you, or you added it intentionally as package specification lines start with two spaces?

Reply:
> So, my new question is it always adding a space for you, or you added it intentionally as package specification lines start with two spaces?

It is intentional, as package specs will always be indented. Should we handle this differently?
At least the docstring of `get_changed_lines` needs a correction.

Reply:
> Should we handle this differently?

It took me a while to understand, so may be we can add a comment. But my only concern is we have no way to ensure this indentation behaviour AFAIK as we are not doing any linting/formatting check on pyproject.toml as of now. Of course during new PR's we can review manually, but it's easy to miss.

Reply:
I've updated the docstring to make this clear. Have a look if that meets your requirement.

Reply:
No more problem on documentation, this is resolved.

Ensuring presence of indentation - should we track or leave it for manual reviews? Either way, out of scope for this PR.

Reply:
> Ensuring presence of indentation - should we track or leave it for manual reviews? Either way, out of scope for this PR.

Is it not automatically ensured by that it's python? If the indentation is missing, it will not compile.

Reply:
I thought indentation was not necessary in TOML for parsing, and package building just zip's the modules, no compilation etc. happens.

![image](https://github.com/sktime/sktime/assets/39331844/9d0474a0-0cc2-4fd3-a9e7-23b987100488)

I removed all indentation in a test branch, and can successfully commit without any complaints from pre-coomit. If you are saying during whl creation step it'll fail, that I am not sure.

Can we check this somehow? Do we have a wheel-creation-without-release CI?

Reply:
what do you suggest?

Reply:
![image](https://github.com/sktime/sktime/assets/39331844/41617337-6b84-4ffa-b01a-b1a43c277af1)

I checked locally and it worked.

Commands used:

1. pip install build
2. python -m build

(2nd command is from repository root directory)

Reply:
> what do you suggest?

Nothing that is automated and foolproof. We can add the pyproject formatter in pre-commit, but that is incompatible with dependabot. So, manual checking is probably the only option.

Another possibility is to write our own hook (shell script or anything) which validates this, but I have no personal experience how to do that.

Reply:
I would go with manual review, and later exploring `dependabot` compatible linters?